### PR TITLE
New Feature: custom Message Element & update parameter for setErrorHTML() & setSuccessHTML()

### PR DESCRIPTION
**Pull request by hungdoan2**
**New feature:**
* Add new function: $validationProvider.addMsgElement()
* Add new function: $validationProvider.getMsgElement()

**Update feature:**
From
````javascript
$validationProvider.setErrorHTML(function(msg){
});

$validationProvider.setSuccessHTML(function(msg){
});
````
To
````javascript
$validationProvider.setErrorHTML(function(msg, element, attrs){
});

$validationProvider.setSuccessHTML(function(msg, element, attrs){
});
````

### CODE REVIEW

**Comment by huei90** [comment]
:+1: This looks good @hungdoan2

File: API.md [hunk 1]
```diff
@@ -360,3 +360,18 @@ angular.module('yourApp', ['validation'])
...
+    .config(['$validationProvider', function ($validationProvider) {        
+        $validationProvider.addMsgElement = function(element) {
+          return $(element).parent().append('<span></span>');
+        };
+        $validationProvider.getMsgElement = function(element) {
+          return $(element).parent().children().last();
+        };
+
+    }]);
+```
```
Comment (huei90) [inline API.md]: Do you mind give more description about `add/getMsgElement` ?

This is difficult to understand,

1. Show the default message element
2. Show the example why we use addMsgElement

File: dist/angular-validation.js [hunk 2]
```diff
@@ -270,6 +270,28 @@
...
 
     /**
+     * Add Message Element at the initial phase
+     * Do this function when to custom your messageElement 
+     * NODE: this funtion have similar purpose with `message-id` attribute. 
+     * This function will help you add your `messageElement` automatically instead of pre-defined.
+     * @param element
+     */
+    this.addMsgElement = function(element) {
+      return element.after('<span></span>');
```
Comment (huei90) [inline dist/angular-validation.js]: Also, I would like to use `set` instead of `add`, as you can see other providers function name, they are all `set` and `get` to name it.

File: API.md (hunk 1 above)
Comment (hungdoan2) [inline API.md]: ok

**Comment by huei90** [comment]
It will be perfect if you can squash your commit into more readable message. :beer:

File: dist/angular-validation.js (hunk 2 above)
Comment (hungdoan2) [inline dist/angular-validation.js]: In my opinion set vs add is absolutely different.
set : `set` value to an existence item/property
add: `add` a not-existence item 
Because Msg Element hadn't existed before I added it => So I use `add` instead of `set`. 
But `set` can be right at some point.

So, I think It's better to merge `addMsgElement` and `setMsgElement` into one `setMsgElement`

In `setMsgElement` we do `add` and return the DOM object of MsgElement. And your directive will get Msg element from returned value from  `setMsgElement`.

**Example:**
````javascript
        $validationProvider.addMsgElement = function(element) {
            $(element).parent().append('<span></span>');
        };
        $validationProvider.getMsgElement = function(element) {
          return $(element).parent().children().last();
        };
````
will be replaced by

````javascript
        $validationProvider.setMsgElement = function(element) {
          $(element).parent().append('<span></span>');
          return $(element).parent().children().last();
        };
````

If you agree with this implement I will update my code.

Comment (huei90) [inline dist/angular-validation.js]: I get your point, using `add` is more understandable in this state - adding a non-existing item.
Then let's stick on the original one `add` :+1:

File: API.md (hunk 1 above)
Comment (huei90) [inline API.md]: Also, I would like to add a line

> Please make sure that you understand the usage before using `add/getMsgElement`

**Comment by huei90** [comment]
@hungdoan2 I'm ready to merge the commit after you finish.

**Comment by hungdoan2** [comment]
@huei90 already squashed all commits & update document.
About document I think It will be easier to follow If there is a navigation bar in document page , which will show all the API/Feature need, like this: http://angular-ui.github.io/ui-router/site/#/api/ui.router

**Comment by huei90** [comment]
Please rebase the master lastest code.

> navigation bar

Yes, this is the ticket I'm working on.

**Comment by hungdoan2** [comment]
@huei90 already rebased :)

File: dist/angular-validation.js [hunk 3]
```diff
@@ -383,12 +407,12 @@
...
 
       if (messageId || validationGroup) messageElem = angular.element(document.querySelector('#' + (messageId || validationGroup)));
-      else messageElem = element.next();
+      else messageElem = $validationProvider.getMsgElement(element);
 
       if (element.attr('no-validation-message')) {
         messageElem.css('display', 'none');
       } else if ($validationProvider.showSuccessMessage && messageToShow) {
-        messageElem.html('').append($compile($validationProvider.getSuccessHTML(messageToShow))(scope));
+        messageElem.html($validationProvider.getSuccessHTML(messageToShow, element, attrs));
```
Comment (huei90) [inline dist/angular-validation.js]: @hungdoan2 This should be rebased too. the $compile.

**Comment by hungdoan2** [comment]
@huei90 hope this commit work :muscle:

**Comment by huei90** [comment]
@hungdoan2 Yes this works perfectly :beers:

**Comment by huei90** [comment]
@lvarayut I'm going to merge this commit, please review.

File: API.md [hunk 4]
```diff
@@ -362,3 +362,43 @@ angular.module('yourApp', ['validation'])
 		};
     }]);
 ```
+
+
+### **Setup Message Element in config phase**
```
Comment (lvarayut) [inline API.md]: Could you explain more in plain English? For example, "You could use this addMsgElement to ... It reduces your effort of manually putting `message-id" so on and so forth.

File: dist/angular-validation.js (hunk 2 above)
Comment (lvarayut) [inline dist/angular-validation.js]: Actually, I agree with @huei90 at the beginning. The name itself doesn't explain what the function does, at least for me. I would prefer `set` over `add`. We don't add anything to the page, however, we set the default place where messages will go. The function might be named as `setDefaultMsgElement` or something alike.

**Comment by lvarayut** [comment]
@hungdoan2 Would you mind to provide some unit tests covering your new functions?

**Comment by huei90** [comment]
@hungdoan2 I can do the unit tests job if you want me to do it.

File: dist/angular-validation.js (hunk 2 above)
Comment (huei90) [inline dist/angular-validation.js]: Sorry, I need to refresh everything.
The reason why I'm choosing `set` for the previous provider is they are allowed user to set something - set the html set the ouput message.

In this `addMsgElement`, this is correct when using within `directive`, because we are adding something not set. Set is the method that allow people to `set`, using `add` to do the set thing in config phase is not that correct (ideally)

@lvarayut @hungdoan2 Let's figure out a better way to use `set` or `add`

**Comment by hungdoan2** [comment]
@huei90 Thanks you, I'm glad to hear that, Would u might helping me with these test cases! . Actually I got a trouble with `grunt test`: It always show `Can not load "PhantomJS", it is not registered!`.  even I installed all phantom packages. I'm checking it now.

**Comment by huei90** [comment]
@hungdoan2 try `npm install`, else `uninstall -g phantomjs` and try install again.

**Comment by hungdoan2** [comment]
@huei90 ok, I will complete my task in this week :muscle: . and I'm sorry to rush you^^

**Comment by huei90** [comment]
@hungdoan2 No worry :smile:

File: test/unit/providerSpec.js [hunk 5]
```diff
@@ -228,4 +228,168 @@ describe('provider', function() {
...
+    it('received correct parameters', inject(function($validation) {
+      $validation.getMsgElement = function() {
+        expect(arguments.length).toEqual(1);
+        expect(angular.isElement(arguments[0])).toBe(true);
+      };
+      $compile('<form name="CustomMsgElmForm"><input type="text" name="required" ng-model="required" validator="required"></form>')($scope);
+
+    })); //END it
+
+  }); //END describe
```
Comment (huei90) [inline test/unit/providerSpec.js]: This is fine :smile: :+1:

Comment (hungdoan2) [inline test/unit/providerSpec.js]: :+1:

**Comment by huei90** [comment]
This commit is almost done, I'm going to merge it now.
